source/ship/ShipAICache.cpp

```diff
@@ -0,0 +1,115 @@
+/* ShipAICache.cpp
+Copyright (c) 2022 by Hurleveur
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ShipAICache.h"
+
+#include "../pi.h"
+#include "../Ship.h"
+
+#include <algorithm>
+#include <cmath>
+
+using namespace std;
+
+
+
+ShipAICache::ShipAICache(const Ship &ship)
+{
+	this->ship = &ship;
+	UpdateWeaponCache();
+}
+
+
+
+void ShipAICache::UpdateWeaponCache()
+{
+	bool hasWeapons = false;
+	bool canFight = false;
+	double totalDPS = 0.;
+	double splashDPS = 0.;
+	double artilleryDPS = 0.;
+
+	shortestRange = 4000.;
+	shortestArtillery = 4000.;
+	minSafeDistance = 0.;
+
+	for(const Hardpoint &hardpoint : ship->Weapons())
+	{
+		const Outfit *weapon = hardpoint.GetOutfit();
+		if(weapon && !hardpoint.IsAntiMissile())
+		{
+			hasWeapons = true;
+			bool lackingAmmo = (weapon->Ammo() && weapon->AmmoUsage() && !ship->OutfitCount(weapon->Ammo()));
+			// Weapons without ammo might as well not exist, so don't even consider them
+			if(lackingAmmo)
+				continue;
+			canFight = true;
+
+			// Calculate the damage per second,
+			// ignoring any special effects. (could be improved to account for those, maybe be based on cost instead)
+			double DPS = (weapon->ShieldDamage() + weapon->HullDamage()
+				+ (weapon->RelativeShieldDamage() * ship->Attributes().Get("shields"))
+				+ (weapon->RelativeHullDamage() * ship->Attributes().Get("hull")))
+				/ weapon->Reload();
+			totalDPS += DPS;
+
+			// Exploding weaponry that can damage this ship requires special consideration.
+			if(weapon->SafeRange())
+			{
+				minSafeDistance = max(weapon->SafeRange(), minSafeDistance);
+				splashDPS += DPS;
+			}
+
+			// The artillery AI should be applied at 1000 pixels range, or 500 if the weapon is homing.
+			double range = weapon->Range();
+			shortestRange = min(range, shortestRange);
+			if(range >= 1000. || (weapon->Homing() && range >= 500.))
+			{
+				shortestArtillery = min(range, shortestArtillery);
+				artilleryDPS += DPS;
+			}
+		}
+	}
+
+	// Calculate this ship's "turning radius"; that is, the smallest circle it
+	// can make while at full speed.
+	double stepsInHalfTurn = 180. / ship->TurnRate();
+	double circumference = stepsInHalfTurn * ship->MaxVelocity();
+	turningRadius = circumference / PI;
```

Comment:
This should be its own caching function separate from updating weapon info. Your turn rate is influenced by your mass, and a lot of things can change a ship's mass, such as dumping cargo or recalling or deploying fighters. Tracking down everywhere where a ship's mass could change and calling the caching function there would be impossible to deal with, though. I actually ran into this situation with caching jump cost information, which may potentially be influenced by the ship's mass. My solution was to have the JumpNavigation class keep track of the last known ship mass it remembers seeing, and then at the start of Engine::MoveShip, recalibrate the ship's jump information if the mass isn't what we last saw.

You may want to peak at the other places where RecalibrateJumpNavigation is called to determine where we might call these caching functions, and it'd probably be better to replace RecalibrateJumpNavigation with a single RecacheInfo function, or something like that.

Reply:
done